git-config/src/file/git_config.rs

```diff
@@ -147,18 +148,57 @@ impl<'event> GitConfig<'event> {
     #[inline]
     pub fn from_paths(paths: &[&Path]) -> Result<Self, ParserOrIoError<'static>> {
         let mut config = Self::new();
-
-        for path in paths {
-            let other = Self::open(path)?;
-            for (section_id, section_header) in other.section_headers {
-                config.push_section(
-                    section_header.name.0.to_owned(),
-                    section_header.subsection_name.to_owned(),
-                    other.sections[&section_id].clone(),
-                );
+        let mut seen = HashSet::new();
+        let paths: Vec<PathBuf> = paths
+            .iter()
+            .map(|p| PathBuf::from(p.to_str().expect("invalid utf-8 path")))
+            .collect();
+        const MAX_DEPTH: u8 = 10;
+
+        fn from_paths_req(
+            paths: Vec<PathBuf>,
+            seen: &mut HashSet<PathBuf>,
+            config: &mut GitConfig,
+            depth: u8,
+        ) -> Result<(), ParserOrIoError<'static>> {
+            if depth >= MAX_DEPTH {
+                return Ok(());
+            }
+            for path in paths.iter() {
+                let other = GitConfig::open(path)?;
+                let mut section_ids: Vec<_> = other.section_headers.keys().collect();
+                section_ids.sort();
+                for section_id in section_ids {
+                    let section_header = &other.section_headers[&section_id];
+                    if section_header.name.0 == "include" {
+                        let path_values = other.sections[&section_id].values(&Key::from("path"));
+                        let mut paths = vec![];
+                        for path_value in path_values {
+                            let interpolated_path = crate::values::Path::from(path_value).interpolate(None).unwrap();
+                            let path = if interpolated_path.is_relative() {
```

Comment:
could nto use `canonicalize()` since the path needs to be relative to the current `path`

Reply:
I still think that canonicalized paths should be possible for use exclusively in the `seen` hashset.